codalab/lib/upload_manager.py

```diff
@@ -1,74 +1,113 @@
 
     def __init__(self, bundle_model, bundle_store):
-        from codalab.lib import zip_util
-
-        # exclude these patterns by default
         self._bundle_model = bundle_model
         self._bundle_store = bundle_store
-        self.zip_util = zip_util
 
-    def upload_to_bundle_store(
-        self, bundle: Bundle, source: Source, git: bool, unpack: bool, use_azure_blob_beta: bool,
-    ):
-        """
-        Uploads contents for the given bundle to the bundle store.
+    @property
+    def storage_type(self):
+        """Returns storage type. Must be one of the values of the StorageType enum."""
+        raise NotImplementedError
 
-        |bundle|: specifies the bundle associated with the contents to upload.
-        |source|: specifies the location of the contents to upload. Each element is
-                   either a URL or a tuple (filename, binary file-like object).
-        |git|: for URLs, whether |source| is a git repo to clone.
-        |unpack|: whether to unpack |source| if it's an archive.
-        |use_azure_blob_beta|: whether to use Azure Blob Storage.
+    def write_git_repo(self, source: str, bundle_path: str):
+        """Clones the git repository indicated by source and uploads it to the path at bundle_path.
+        Args:
+            source (str): source to git repository.
+            bundle_path (str): Output bundle path.
+        """
+        raise NotImplementedError
 
-        Exceptions:
-        - If |git|, then the bundle contains the result of running 'git clone |source|'
-        - If |unpack| is True or a source is an archive (zip, tar.gz, etc.), then unpack the source.
+    def write_fileobj(
+        self, source_ext: str, source_fileobj: IO[bytes], bundle_path: str, unpack_archive: bool
+    ):
+        """Writes fileobj indicated, unpacks if specified, and uploads it to the path at bundle_path.
+        Args:
+            source_ext (str): File extension of the source to write.
+            source_fileobj (str): Fileobj of the source to write.
+            bundle_path (str): Output bundle path.
+            unpack_archive (bool): Whether fileobj is an archive that should be unpacked.
         """
-        bundle_path = self._bundle_store.get_bundle_location(bundle.uuid)
+        raise NotImplementedError
+
+    def upload_to_bundle_store(self, bundle: Bundle, source: Source, git: bool, unpack: bool):
+        """Uploads the given source to the bundle store.
+        Given arguments are the same as UploadManager.upload_to_bundle_store()"""
         try:
+            # bundle_path = self._bundle_store.get_bundle_location(bundle.uuid)
             is_url, is_fileobj, filename = self._interpret_source(source)
             if is_url:
                 assert isinstance(source, str)
                 if git:
-                    file_util.git_clone(source, bundle_path)
+                    bundle_path = self._update_and_get_bundle_location(bundle, is_directory=True)
```

Comment:
`is_directory` is true because all git repos are directories?

Reply:
Yeah